### ISSUE TITLE
Delayed actions processed in wrong order

### ISSUE THREAD

**Issue by bri365** [StackStorm/st2#2797]
When actions are delayed for policy concurrency (and concurrency.attr), they appear to be retried in newest to oldest order, rather than oldest to newest.  The file attached is a trimmed down actionrunner log showing delayed actions and their subsequent dispatch.  When multiple actions are delayed, they appear to be dispatched in the opposite order intended.

[2600.txt](https://github.com/StackStorm/st2/files/352933/2600.txt)

st2actions/policy/concurrency.py:88 and concurrency_by_attr.py:115 both include order_by['start_timestamp'], but does not indicate ascending or descending.

**Comment by Kami** [comment]
Thanks for the report.

We don't officially say that we process delayed executions in the FIFO order, but I agree with you that LIFO behavior is confusing and making it FIFO would indeed be better and make more sense.

As far as the actual code / implementation goes - I believe the default mongoengine order (unless explicitly specified) is to sort in ascending order so there must be something else going on and I will dig into the code.

**Comment by Kami** [comment]
I quickly looked at the code and the concurrency stuff looks correct - we don't explicitly specify ASC order, but that's the default behavior (so doing `+start_timestamp` is optional).

There might be an issue with the way requested and scheduled are processed, but looking at the code nothing stands out. By default RabbitMQ messages are processes in FIFO order, but there are many edge-cases (in a distributed system) where this doesn't always hold true - if there are multiple consumers (we run multiple action runners by default), if for some reason message is re-queued (timeout, ...), etc.

Do you happen to have a simple workflow / action example which I can use to try to reproduce this issue?

/cc @m4dcoder

**Comment by bri365** [comment]
Thanks, Kami.  I'll try to see if we can reproduce this issue.  Also, our system is defaulting to a single action runner (I think four cores or less yields a single one) - any known concerns?  Would we be better/safer with two?

**Comment by Kami** [comment]
No, it should be fine with a single action runner - if anything it should be better with regards to the race conditions (but there shouldn't be any, more runners simply means more actions will be executed in parallel).

And yeah, if you can come up with a simple workflow with reproduces a problem, please gist it here and I will have a look.

**Comment by m4dcoder** [comment]
@Kami, nope, will have to define one to test this.

**Comment by bri365** [comment]
I'm working on a simple set of repro steps for this and encountered another issue. I can't disable a policy!  I set 'enabled' to false and reload and the policy is still enforced.  I remove the file, st2ctl stop/start, reboot, and the policy is still being applied!  I feel a little silly asking, but is there extra special policy persistence somewhere?

**Comment by Kami** [comment]
@bri365 It could be a bug - I will look into it.

**Comment by Kami** [comment]
@bri365 Yeah - it looks like a bug in the code.

I'm working on a fix as we speak.

**Comment by bri365** [comment]
@Kami @m4dcoder after developing and verifying repro steps for my issue, I have isolated why items appeared "out of order" to me and in the process found another one or two issues, which I will file separately.

In this case, my clean-up script to delete stuck actions included iterating through the list of 'scheduled', 'delayed', and 'running' actions and canceling them.  The iteration was based on 'st2 execution list -l --status scheduled -n 1' which delivers the most recent action in the requested state, rather than the oldest action of the given state.

I think it is valuable to get the oldest action/workflow in a given state, so either reversing the order of returned items or offering a sort order would be very helpful.

**Comment by bri365** [comment]
Closing this as not an issue and opening a feature request for execution list sort order.
